ceph: only prepare partitions if gpt (#8200)
rook/rook: Documentation/ceph-common-issues.md

>unable to recognize "STDIN": no matches for kind "CustomResourceDefinition" in version "apiextensions.k8s.io/v1"
>```
You need to create the CRDs found in `cluster/examples/kubernetes/ceph/pre-k8s-1.16`. Note that these pre-1.16 `apiextensions.k8s.io/v1beta1` CRDs are deprecated in k8s v1.16 and will no longer be supported from k8s v1.22.

## Unexpected partitions created

### Symptoms

> **Rook versions v1.6.0-v1.6.6 may create unwanted OSDs on these partitions, which will corrupt

Reviewer: "these partitions" --> which partitions? since it moved up to the top, maybe needs rewording.

existing OSDs.**

Unexpected partitions are created on host disks that are used by Ceph OSDs. This happens more often
on SSDs than HDDs and usually only on disks that are 1TB or larger. Many tools like `lsblk`,
`blkid`, and `udevadm` will not show a partition table type for the partition; however, `parted`
will reliably list the partition as `atari` (see the example command and output below).

```console
# parted --script /dev/sdb1 print
Model: Unknown (unknown)
Disk /dev/sdb1: 1509kB
Sector size (logical/physical): 512B/512B
Partition Table: atari
Disk Flags:

Number  Start  End  Size  Type  File system  Flags

```

The underlying issue causing this is Atari (sometimes identified as AHDI) partition support in the
Linux kernel. Atari partitions have very relaxed specifications compared to modern partition types,
and it is relatively easy for random data written to a disk to appear as an Atari partition to the
Linux kernel. Ceph's Bluestore OSDs have an anecdotally high probability of writing data on to disks
that can randomly appear to the kernel as an Atari partition.

You can see https://github.com/rook/rook/issues/7940 for more detailed information and discussion.

### Solution

> If you are using Rook v1.6, you must update to v1.6.7 or higher to avoid OSD corruption.

Reviewer: Will an extra OSD be created on the smaller partition that needs to be cleaned up? Or is there no action item besides upgrading?

Author: If an OSD is already created on the smaller partition, I think the only recourse is to wipe the disk and deploy again because there is already corruption.

Author: Updated the doc with an example recovery.

If you are using Rook v1.6.7 or higher, it is safe to ignore these unexpected `atari` partitions.
Rook will only deploy OSDs on `gpt` partitions.

Not all Linux distributions build kernel with Atari support enabled. If possible, we recommend

> switching to a kernel built without Atari support enabled. The example below shows an example of how

Reviewer 2: ```suggestion
switching to a kernel built without Atari support enabled. The example below shows of how
```

to determine if your kernel was built with Atari support enabled.

```console

> $ cat /boot/config-$(uname -r) | grep ATARI

Reviewer 2: ```suggestion
$  grep ATARI /boot/config-$(uname -r)
```

### Symptoms
Unexpected partitions are created on host disks that are used by Ceph OSDs. This happens more often
on SSDs than HDDs and usually only on disks that are 1TB or larger. Many tools like `lsblk`,
`blkid`, and `udevadm` will not show a partition table type for the partition; however, `parted`
will reliably list the partition as `atari` (see the example command and output below).

```console
# parted --script /dev/sdb1 print
Model: Unknown (unknown)
Disk /dev/sdb1: 1509kB
Sector size (logical/physical): 512B/512B
Partition Table: atari
Disk Flags:

Number  Start  End  Size  Type  File system  Flags

```

The underlying issue causing this is Atari (sometimes identified as AHDI) partition support in the
Linux kernel. Atari partitions have very relaxed specifications compared to modern partition types,
and it is relatively easy for random data written to a disk to appear as an Atari partition to the
Linux kernel. Ceph's Bluestore OSDs have an anecdotally high probability of writing data on to disks
that can randomly appear to the kernel as an Atari partition.

You can see https://github.com/rook/rook/issues/7940 for more detailed information and discussion.

> **Rook versions v1.6.0-v1.6.6 may create unwanted OSDs on these partitions, which will corrupt

Reviewer: How about mentioning at the top of this topic that this issue may affect these versions?

> existing OSDs.**

Reviewer: The corruption comes from the OSD being created on the smaller 48Gi partition, right? We should probably release 1.6.7 next week to get this fix out there instead of waiting for the usual cycle.

Reviewer: After we get this merged and before the v1.6.7, it would be helpful if someone hitting this issue could confirm the fix in their env.

Author: > The corruption comes from the OSD being created on the smaller 48Gi partition, right? We should probably release 1.6.7 next week to get this fix out there instead of waiting for the usual cycle.

Yes. I agree.

### Solution
If you are using Rook v1.6, you must update to v1.6.7 or higher to avoid OSD corruption.

If you are using Rook v1.6.7 or higher, it is safe to ignore these unexpected `atari` partitions.
Rook will only deploy OSDs on `gpt` partitions.

Not all Linux distributions build kernel with Atari support enabled. If possible, we recommend
switching to a kernel built without Atari support enabled. The example below shows some examples of
how to determine if your kernel was built with Atari support enabled.

```console

> $ cat /boot/config-4.15.0-144-generic | grep ATARI

Reviewer 3: Sometimes the /boot directory can hold many configs, with the example below we can avoid misunderstanding   : 

```bash
cat /boot/config-$(uname -r) | grep ATARI
```

CONFIG_ATARI_PARTITION=y

> $ zcat /proc/config.gz | grep ATARI

Reviewer 2: ```suggestion
$ zgrep ATARI /proc/config.gz
```

Also on F31 (and above I believe), this is not needed and the file does not exist. This seems sufficient:

```sh
grep ATARI /boot/config-$(uname -r)
# CONFIG_ATARI_PARTITION is not set
```

Sector size (logical/physical): 512B/512B
Partition Table: atari
Disk Flags:

Number  Start  End  Size  Type  File system  Flags

```

The underlying issue causing this is Atari (sometimes identified as AHDI) partition support in the
Linux kernel. Atari partitions have very relaxed specifications compared to modern partition types,
and it is relatively easy for random data written to a disk to appear as an Atari partition to the
Linux kernel. Ceph's Bluestore OSDs have an anecdotally high probability of writing data on to disks
that can appear to the kernel as an Atari partition.

You can see https://github.com/rook/rook/issues/7940 for more detailed information and discussion.

### Solution
#### Recover from corruption (v1.6.0-v1.6.6)
If you are using Rook v1.6, you must update to v1.6.7 or higher to avoid further incidents of OSD
corruption caused by these Atari partitions.

If there are already OSDs created on the unexpected partitions, OSD corruption has already occurred.
To recover from this, immediately update to v1.6.7 or higher. After the update, no further OSD
corruption should occur. Next, to get back to a healthy Ceph cluster state, focus on one corruped

> disk at a time and remove all OSDs on each corrupted disk one disk.

Reviewer: A link to this topic may be helpful
```suggestion
disk at a time and [remove all OSDs](ceph-osd-mgmt.md#remove-an-osd) on each corrupted disk.
```

Author: There is a link later in the doc.
